### ISSUE TITLE
Async

### ISSUE THREAD

**Issue by littledan** [proposal-signals/proposal-signals#30]
Lots of people here are interested in handling async/await in conjunction with signals. Some approaches that have been suggested:
- @modderme123 's Bubble reactivity uses something like React Suspense, where a pending promise state can be propagated through the dependency graph. This might require support from the core for efficiency of propagating that information and to fully eliminate the "function coloring problem".
- @pzuraq has proposed "Relays" to connect graphs together in a way that may be managed in a naturally async way. This may benefit from built-in support as well--it can mostly be implemented using effects, but ownership would be different.
- @alxhub has wondered whether it would be possible to make async/await in a computed "just work", but also doesn't consider this an MVP requirement.

**Comment by lgarron** [comment]
I know it has already been decided to exclude this from an initial version of the proposal, but I think it would be *immensely* valuable to have something forward-compatible with an async mechanism, because:

- Web page performance benefits greatly from loading as little JS as possible.
- Using dynamic imports is a well-established mechanism to load heavier JS incrementally as it's needed.
- Similarly, it is often necessary to load data as needed, because sending all the data ahead of time is prohibitively expensive or literally impossible.
- For any calculations that take over ≈10ms, it is critical to offload the calculations to web workers. This requires the usage of `async`/callback code.
  - It is possible to make sync calls to WASM, but if the WASM code is not guaranteed to be loaded ahead of time then loading WASM is a similar issue to dynamic code import.

While there is a case to be made that app developers can solve this by manually implementing "loading" signals/states or carefully interleaving any async setup work, it can be much more straightforward to wait for the code/data to be fetched. In many cases, the user would expect to wait a brief while for an operation to finish, and it's still possible to add loading states to an async signal network.

I maintain a library named [`cubing.js`](https://github.com/cubing/cubing.js) for displaying puzzles. We need to:

- stick with compatible web technologies — no framework lock-in,
- support a wide variety of puzzles (too much to load all at once), and
- efficiently recompute puzzle state across [millions of moves](https://experiments.cubing.net/cubing.js/stress-tests/2x2x2-devils-alg.html).

After spending a lot of time looking at options, I ended up writing a small mechanism that propagates signals across a DAG: https://github.com/cubing/cubing.js/blob/cdfab5cb35c3741ef80e0680d9b72c69263205d3/src/cubing/twisty/model/props/TwistyProp.ts

I know it can't be a universal solution, but it has done a great job of being performant, avoiding synchronization bugs, and getting most of the benefits of async code while supporting `async` nodes. In particular, *each node receives a fully awaited set of inputs from parents* at computation time. Here is a simplified explanation of the algorithm:

<details>

There is a global `globalSourceGeneration` counter.

Each node has:

- The following construction-time inputs:
  - A [`Record`](https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeys-type) `parents` of parent nodes (passed at construction time). Each parent has a `parentKey` that is used to index other `Record`s in the algorithm.
  - A synchronous function `canReuseValue(v1, v2)`, defaulting to a `===` comparison. (This could probably support async.)
  - An async function `derive(inputs)`, defaulting to a `===` comparison.
- A set of `childNodes` (populated by children when they are constructed).
- A `Record` of `cachedInputs` (one field corresponding to each parent).
- A `lastSourceGeneration` number.
- A `lastComputedGeneration` number.
- A `cachedValue`.
  - This value is "up to date" when `lastSourceGeneration` == `lastComputedGeneration`.

When a node is updated:

- Increment `globalSourceGeneration`.
- *Synchronously* traverse all descendants and mark them as stale by setting their `lastSourceGeneration` fields to `globalSourceGeneration`.

When someone calls `node.get()`:

- If `lastSourceGeneration == lastComputedGeneration`, return `cachedValue`, else:
- Call `parents[parentKey].get()` for all parents in parallel (using `Promise.all`) and store the in `newInputs` by `parentKey`.
  - If `parents[parentKey].canReuseValue(cachedInputs[parentKey], newInputs[parentKey]) == false` for any parent:
    - Set `cachedValue = await derive(cachedInputs)`[^1].
- Update `newInputs = cachedInputs` and `lastComputedGeneration = lastComputedGeneration`.
 - Return `cachedValue`.

[^1]: Optimization: if `cachedValue` has not changed (based on `canReuseValue`), reuse the existing cached value so the new one can be GCed.

To receive updates of the freshest value for a given `node` as soon as possible: When a node is marked as stale, schedule `node.get()` call using `queueMicrotask(…)` and dispatch updates with the results.[^2]

[^2]: Alternatively, perform this work in the same microtask by building a list of these scheduled calls while marking nodes as stale, and call them all once the graph has been completely traversed.

There are some details around mutability rules and deduplication optimizations I've skipped over, but I think this captures enough core details.
</details>

**Comment by NullVoxPopuli** [comment]
@lgarron I'm making a utility library here: https://github.com/NullVoxPopuli/signal-utils/pull/1 which will include a reactive async implementation built upon the polyfill (and will eventually phase out the polyfill when real implementation occurs).

would that help with your concerns?

**Comment by lgarron** [comment]
> @lgarron I'm making a utility library here: [NullVoxPopuli/signal-utils#1](https://github.com/NullVoxPopuli/signal-utils/pull/1) which will include a reactive async implementation built upon the polyfill (and will eventually phase out the polyfill when real implementation occurs).
> 
> would that help with your concerns?

Hmm, the main appeal of the Signals proposal is that the code would be built-in and standardized.

I support a light async layer on top of the existing proposed spec could be handy, but using any library has costs and liabilities beyond using standard web platform code. In particular, if Signals becomes a standard for interoperable signaling code, then any async implementations on top of that would need to be careful to interoperate, probably to the point that… we would need a specification for now to interoperate. 😆

**Comment by timonkrebs** [comment]
> pending promise state can be propagated through the dependency graph

Not only the promise state would be of interest. There will be a need to cancel all pending evaluations. Eor example if you switch the whole page and need to cancel all the requests that are still pending.

> When you work with a system like this, you end up wanting a signals-land version of Promise.all to prevent a computation from rerunning multiple times

What signals need is [Structured Concurrency](https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/). I am already looking at this with: https://github.com/timonkrebs/MemoizR

**Comment by lgarron** [comment]
> but since a lot of things can be implemented with the currently described primitives, deferring decisions to future proposals via libraries feels like the best way to move the existing proposal forward

I also don't see a way to make deduplication work with `Promise`s passed around a sync `Signals` API, since the definitions require `equals(...)` to be implemented as synchronous. I may be missing something, though.

**Comment by pzuraq** [comment]
@lgarron the way that you could handle that is via what I've been calling the "Relay" pattern, mentioned by @littledan above. I'm going to write up a document describing this later this week if I can find some time, but here's an example:

```ts
// $lib/data-utils.ts
import { client } from '$lib/juno';

export const fetchJson = (url: Signal<string>) => {
  let state: Signal.State;
  let controller: AbortController;

  const loadData = new Signal.Computed(async () => {
    controller?.abort();

    // Update isLoading to true, but keep the rest of the state
    state.set({ ...state.get(), isLoading: true });

    controller = new AbortController();
    const response = await fetch(url.get(), { signal: controller.signal });
    const value = await response.json();

    state.set({ ...state.get(), value, isLoading: false });
  });
  
  return new Signal.Computed(() => {
    state = new State({ isLoading: true, value: undefined });   
    
    loadData.get();

    return state;
  }, {
    onObserved() {
      loadData.get();
    },

    onUnobserved() {
      controller?.abort();
    }
  });
}
```

The basic idea is that Relays absorb the async of a subgraph and expose it to a parent graph that is consuming the subgraph in a synchronous way. So, you can intercept those sync changes and watch for equality at that point.

**Comment by dead-claudia** [comment]
So, I figured out the minimal primitive needed to make async signals work without needing to deal with async functions directly: an "is updating" condition propagated similarly to the "is dirty" condition underlying `Watcher.prototype.getPending`.

- `Signal.State` would allow this to be set directly, as part of setting the value. It should be optional, defaulting ideally to `false`.
- This condition can be read directly from any signal. For `Signal.State`, it can return the value of itself. For `Signal.Computed`, it should return `true` iff all its source signals return `true` for the value.
- `Signal.State` needs a way to store exceptions like `Signal.Computed`.
- Reads of this condition must be auto-tracked.

Then, you can implement async `Computed`s in userland as follows:

```js
// The `.set` options are purely hypothetical here
function asyncComputed(body, opts) {
    const state = new Signal.State()
    let token = {}

    const invoker = new Signal.Computed(async () => {
        const prev = Signal.subtle.untrack(() => state.get())
        state.set(prev, {isPending: true})

        const t = token = {}
        let failed = false
        let value
        try {
            value = await body()
        } catch (e) {
            value = e
            failed = true
        }

        // Signal refreshed, drop the stale value
        if (t !== token) {
            if (failed) throw value
            return
        }

        state.set(value, {
            isException: failed,
            isPending: false,
        })
    })

    return new Signal.Computed(() => {
        invoker.get()
        return state.get()
    }, opts)
}
```

When you use it, it literally works like sync signals in every way, except that those who care can transparently know if a given value is ready. No inner function coloring involved.

```
const state = getStateSomehow()

effect(() => {
    if (state.isPending) showLoadingSymbol()
    updateUI(state.get())
})
```

**Comment by NullVoxPopuli** [comment]
@dead-claudia exactly! That's what i did here:
- https://github.com/NullVoxPopuli/signal-utils?tab=readme-ov-file#async-function
- https://github.com/NullVoxPopuli/signal-utils/blob/main/src/async-function.ts

**Comment by dead-claudia** [comment]
@NullVoxPopuli That doesn't seem to propagate readiness through computeds? Or am I missing something?

That propagation is pretty crucial to avoiding the function coloring problem in signals.

**Comment by NullVoxPopuli** [comment]
What's the tldr on the coloring problem?

Here is a test that shows behavior, if that helps: https://github.com/NullVoxPopuli/signal-utils/blob/ef3d29f2dd2ff943714c5f7c1bfd0c89af529ad7/tests/async-function.test.ts#L111

State changes are propagated appropriately (and non duplicatively) when tracked states are consumed by the outer computed.

But maybe the coloring problem will inform me what needs fixed 🤔

**Comment by dead-claudia** [comment]
@NullVoxPopuli Okay, maybe "function coloring" was a bit imprecise. I'm also conflating it with "setting to a thrown exception" and so I'll drop that for now.

It's more so "value coloring" that's at play here.

Without an "is pending" flag, you have to model your signal as follows:

```ts
type AsyncValue<T> =
    | {thrown: false, value: T, isPending: boolean}
    | {thrown: false, value: Error, isPending: boolean}
```

That `isPending` boolean allows that declarative branching I mentioned above.

Now, suppose you want to filter the list by type. If you have a sync collection of records built by the user, and an async collection of records fetched from the network, you'll need two different functions to collect them:

> Obviously this is academic, but not all signal combinators can just be turned into simple functions like that.

```js
function filterSync(type, source) {
    return new Signal.Computed(() => {
        const t = type.get()
        return source.get().filter((s) => s.type === t)
    })
}

function filterAsync(type, source) {
    return new Signal.Computed(() => {
        const t = type.get()
        const result = source.get()
        if (result.thrown) return result
        return {
            ...result,
            value: result.filter((s) => s.type === t),
        }
    })
}
```

Computed signals already have the ability to store exceptions, so those can be leveraged accordingly.

```js
class StateWithException extends Signal.Computed {
    #state

    constructor(initial, opts) {
        const state = new Signal.State({thrown: false, value: initial})
        super(() => {
            const {thrown, value} = state.get()
            if (thrown) throw value
            return value
        })
    }

    setException(e) {
        this.#state.set({thrown: true, value: e})
    }

    set(v) {
        this.#state.set({thrown: false, value: v})
    }
}
```

> Obviously, that `StateWithException` is inefficient. This is where my suggestion of, maybe, a `signal.setException(...)` would come in handy.

However, this only eliminates the need to propagate one of the fields. We still have another field that needs propagated, and it needs propagated for errors as well as values.

```ts
type AsyncValue<T> = {value: T, isPending: boolean}
type AsyncError = {value: Error, isPending: boolean}
```

Adding an `isPending` value that's propagated across the graph will solve this value coloring problem. And so you can use the same combinators for both sync and async signals, without needing to treat them differently.

```js
function filter(type, source) {
    return new Signal.Computed(() => {
        const t = type.get()
        return source.get().filter((s) => s.type === t)
    })
}
```

Likewise, you can switch from async to sync without code change, and it'll just work.

Suppose somewhere down the line, you make some changes to `getStateSomehow` and `currentDetail` now gets all its info right away from a cache and is no longer async. You're also anticipating a new feature that'll mean `lastEvent` now depends on a network fetch, but your test endpoint isn't up yet.

Then, a month later, that test endpoint comes live and you need to make that change to `lastEvent` in preparation for that feature launch.

> I've made a similar sequence of changes myself in promise-based code before, with a few months in between. And with promises, that's two pretty heavy changes in the view code.

The beauty of this approach is that your view caller in this case won't need rewritten or even changed at all for either case. You can even keep the `if (signal.isPending)` branch in anticipation for your `lastEvent` change. You get maximum freedom here to handle it how *you* want.

**Comment by NullVoxPopuli** [comment]
Ah ok, thanks for explaining! This is indeed a familiar concept than i've been trying to find ergonomics for, and one that @ef4 has been pushing back on @wycats and I to figure out with "Resources", which are like a computed, but with a lifetime and cleanup.

Also! This reminds me a lot of railway-oriented-programming, (https://fsharpforfunandprofit.com/rop/), or Result types in other languages, where, if you can force the whole system to adhere to the same interface at each interaction, there are far fewer surprises.

This is super important for composing multiple computeds represents async state... Without a shared interface at each layer, making sure the pending and error states are appropriately propagated out to the renderer can be cumbersome and error prone 🤔

**Comment by dead-claudia** [comment]
@NullVoxPopuli Admittedly, that style of programming was of *some* inspiration for my idea there, though I wasn't specifically thinking of it by name. (I have heard of the concept before, though.) Also, I was inspired by promises in that they also have this (implicit and internal) "pending" vs "settled" state, and I knew that exposing this data path would allow me to use that paradigm.

I also come from a *huge* virtual DOM background, both as a user and as a framework implementor, and that kind of branching-on-state reigns supreme in render functions/methods. I also have a little bit of recent experience in 2D rendering. So it was only natural for me to think about branching like that. 🙂

**Comment by shaylew** [comment]
@dead-claudia Hah, yeah, this is almost exactly the core construction of the [bubble-reactivity](https://github.com/bubblegroup/bubble-reactivity/blob/main/src/index.ts) approach that @modderme123 and I were fleshing out a while back, and which I tried to describe (maybe poorly) earlier in this thread. The core idea is that loading-ness propagates contagiously forwards on `get`, plus states letting you set their error/loading flags in addition to their value; the rest of the stuff I was talking about earlier is kind of orthogonal to that core.

"Every signal can be loading" avoids the function coloring problem in exactly the same way JS exceptions avoid the coloring problem you'd get by having things return `Result`: by making the ability to be loading implicit everywhere, it doesn't have to be announced or managed explicitly. You're basically trading off between an explicit monadic approach (Promise, Result, Loadable) that shows up in the return type, and an implicit side effects based approach (`throw`, loading state propagation) where some effects happen directly, without being part of the returned value.

As far as side effects go, this is about as benign as you can get -- the side channel is just one bit that gets captured alongside the result, with no control flow implications (unlike `throw`). I think it's a neat trick, and that it's far and away more pleasant to use than having to make things explicitly a `Signal<Loadable<T>>` all over the place.

IIRC it's not a completely satisfying solution to async operations for all frameworks, because some framework want to let you write  "effects that wait for things to load before running" without you having to manually check that things are loaded or perhaps without you having to see the default unloaded values of signals show up in their types at all. OTOH I don't know if anyone has any alternatives that actually deliver on that experience for effects without introducing Weird Stuff:tm: (throw-and-restart based suspense, and/or spreading what should be one effect across multiple microtasks because you have to `await` something)... so it's not clear that not knowing what to do about async effects should keep us from having this rather well-behaved system for async computeds.

**Comment by dead-claudia** [comment]
Related: https://github.com/tc39/proposal-signals/issues/178

**Comment by pzuraq** [comment]
I finally got around to my writeup on the Relay pattern, you can check it out here: https://www.pzuraq.com/blog/on-signal-relays

My current feeling is that Relays would provide a primitive that would work well with all forms of async, including:

- Request/response (and promises in general)
- Pub/sub
- Background threads
- Websockets

Basically, they can handle 1-to-1 connections AND 1-to-many connections quite well in a variety of ways.

At the moment, I'm concerned about over optimizing for 1-to-1/promise oriented async with automatic propagation of `isPending` and error states. I do believe that is the _majority_ of async that exists out there, so I can definitely see adding more features specifically for those cases to reduce boilerplate, but I think the two things we should be designing for are:

1. We shouldn't actively _prevent_ modeling of non-promise based async. If the default path is to use a promise and propagate automatically that is great, but it should still be _possible_ to add signals that can handle other forms of async and incorporate them into output.
2. We should either be very sure that the propagated values (`isPending`, `error`, etc) are the only values that should ever be propagated, OR that the mechanism for propagating them is pluggable/extensible by future proposals. If there's a reasonable possibility of adding more types of values that are automatically propagated (e.g. some sort of automatic propagation for pub/sub state? Unsure what that would look like) I would lean toward the later, but _so far_ I can't think of anything.

**Comment by rjgotten** [comment]
Has anyone yet considered the [async context proposal](https://github.com/tc39/proposal-async-context) ?

Where basically an async context could track dependencies of derived/computed signals and an API set up along those routes would allow callers to able to do something like the following totally transparently, _'and it just works.'_

All the way down into any async function tapping any other observables.

```js
Signal.derivedAsync(async abortSignal => {
  const result = await fetch(url.get(), { signal: abortSignal });
  return await processAsync({ result, dependency: otherDependency.get(), signal: abortSignal });
}
```

**Comment by EisenbergEffect** [comment]
@rjgotten Fortunately, @littledan is involved both with async context and signals. I think we would like to integrate them. Dan can probably provide some additional thoughts.

**Comment by littledan** [comment]
My thoughts:
- On Signal.derivedAsync: This is an important space that should have a solution someday, but I don't know how it should look. I don't think just leaving a signal in the "computing" state serves all the use cases--you usually want to get at the previous value or a placeholder, rather than an exception. In general, autotracking is currently a very synchronous mechanism and not suitable for modeling by AsyncContext.
- My initial intuition has been that AsyncContext should tie into Signal.Computed by treating this as a "registration time" case--usually, you should restore the AsyncContext variables when running the computed body, e.g. if you use one to track the current cancel token. However, as @shicks has pointed out, sometimes you actually do want to trace the other path, which is what actually caused the computed to be forced... I'm not sure what the overall solution is.

**Comment by dead-claudia** [comment]
@rjgotten To add on to that, signal state is currently (at least in this repo) specified to use an internal async context variable for glue.

Not to imply this will continue to be the case, just that it currently *is* the case.

**Comment by rjgotten** [comment]
> @littledan 
> On Signal.derivedAsync: This is an important space that should have a solution someday, but I don't know how it should look. I don't think just leaving a signal in the "computing" state serves all the use cases--you usually want to get at the previous value or a placeholder, rather than an exception. In general, autotracking is currently a very synchronous mechanism and not suitable for modeling by AsyncContext.

This is resolveable if `derivedAsync` returns a promise as the result, rather than try to hold on to last computed value while it is internally 're-computing' and also tracks the state of said promise.

It wouldn't be problematic wrt auto-tracking either.
The AsyncContext flowing alongside the async delegate's execution would collect dependents in a collection that is added onto as the delegate advances, and `derivedAsync` would be observing for any changes to those dependents.

If an update were needed mid-execution (i.e. while the promise is not yet settled) then internally the delegate should re-execute. While _externally_ what's observed could still remain the same pending promise. (Only when a promise has already settled would the `derivedAsync` itself need to change value to a new pending promise)

When such a mid-execution update would be needed, the old monitoring/recording in the AsyncContext should be marked as dead - not to be responded to further - so that any dependents still being added or changing in the remainder of the old execution, will no longer cause changes. And then the delegate should be re-executed with a new AsyncContext.

An abort signal would probably also be passed as a parameter to the userland async delegate for cooperative cancellation of the old execution, which might be beneficial to avoid continuing expensive recomputing; to avoid no longer needed network requests; etc.

Then create additional helpers that take care of unwrapping the asynchronous promise into an observable signal's value change. E.g. if you'd want to just expose the promise's state in full:

```ts
Signal.fromPromise<T>(signal: Signal<Promise<T>): { value : T, error : unknown, state : "resolved"|"rejected"|"pending" };
```
And from there, other variants such as holding onto the last known value or error are also possible.



Basically; this means `derivedAsync` actually returns a result synchronously - namely: the promise. And the only thing actually asynchronous about the signal, is that internal AsyncContext carrying the machinery for dependents tracking. That is: the dependents may still mutate even after the signal has already returned a value (a new pending promise). So any solutions like graph coloring that are need to order the dependency graph of pending updates and perform them correctly and efficiently, would need to be aware of that. _However_ they would basically be looking at a synchronous snapshot of the dependents anyway, afaict, and they would need to finish their work synchronously. So there's no problem with dependents changing _while_ they're doing their work, anyway...

**Comment by dead-claudia** [comment]
Okay, I've recently through some experimentation come up with a slightly different model that IMHO is useful in its own right: `Signal.try(func, ...args)`.

> This isn't meant to replace [my `isPending` suggestion](https://github.com/tc39/proposal-signals/issues/30#issuecomment-2050803875), but to augment it.

What this does is run a block, catching errors and indirectly capturing pending signal status. It returns the same kind of object that `Promise.allSettled` returns inside its result array, but with an extra `isPending` boolean that's set to `true` if any inner accessed signal is marked as pending, `false` otherwise.

This also delimits calls: signal accesses in child `Signal.try` calls are ignored by parent `Signal.try` calls.

Note that this shouldn't use async context, only global state, as that limits object leakage and it's a lot cheaper.

**Comment by justinfagnani** [comment]
Just coming to this discussion now, but I hit some of the needs described here when using signals to model data structures that have async edges, ie. to lazy-loaded nodes in the graph or expensive computations.

What I've made personally for this is a class called AsyncComputed, that works like this:
- It takes a Promise-returning computation function with an AbortSignal argument: `(signal: AbortSignal) => Promise<T>`
- The function is called within an internal Signal.Computed, so all synchronous signal access is tracked
- When the internal computed runs, it re-runs the compute function, pre-empting any pending run and aborting their abort signals.
- It has signal-backed getters for its state:
  - `state`: one of `'initial'`, `'pending'`, `'complete'`, or `'error'`. `state` will _synchronously_ change to `'pending'` when dependency signals are dirtied.
  - `value`: the last completion value if the last completion type was `'complete'`, otherwise undefined
  - `error`: the last error if the last completion type was `'error'`, otherwise undefined
- It has a `complete` getter that returns a Promise that settles when the AsyncComputed does. If a new run is triggered before the previous completes, the same Promise instance is kept. If a new run is started after a completion, a new Promise is created.
- It has a `get()` method that returns the latest completion value or throws if the last completion was an error.
- It has a `run()` method to manually read from the internal computed and trigger a run. Otherwise the computed is read from the various getters.

This is somewhat similar to `derivedAsync()` above, Relays, or `signalFunction` from signal-utils - I think the differences are relatively minor. It's more-or-less a subclass of Computed that has extra fields to represent the state, and where `get()` returns the previous completion value while pending.

It's working well enough for me so far, that the only thing I think I'm missing from the proposal is tracking signal reads after the first await of the compute function. If would be really great if computeds over async functions could "just work", although that seems difficult to polyfill.

Propagating pending state through the `value`, and `error` signals is interesting, but I'm unsure how necessary it is. Chaining AsyncComputeds can rely on `state` and `completed` to convey the async pending state and wait for completion.

Chaining a regular Computed off an AsyncComputed just works.

```ts
function filterSync(type, source) {
    return new Signal.Computed(() => {
        const t = type.get();
        return source.get().filter((s) => s.type === t);
    });
}
const source = new AsyncComputed(() => getItems());
```

If `source` here is an AsyncComputed, then `source.get()` will return or throw as it updates, causing the computed to return or throw.

<details>
<summary>My test for chaining</summary>

```ts
  test('chaining a computed signal propagates error state', async () => {
    const dep = new Signal.State('a');
    const task = new AsyncComputed(async () => {
      // Read dependencies before first await
      const value = dep.get();
      await 0;
      if (value === 'b') {
        throw new Error('b');
      }
      return value;
    });
    const computed = new Signal.Computed(() => task.get());
    assert.strictEqual(computed.get(), undefined);

    await task.complete;
    assert.strictEqual(computed.get(), 'a');

    dep.set('b');
    await task.complete.catch(() => {});
    assert.throws(() => computed.get());

    dep.set('c');
    await task.complete;
    assert.strictEqual(computed.get(), 'c');
  });
  ```
  
  </details>

**Comment by justinfagnani** [comment]
@eKoopmans AsyncComputed does have a `status` signal that includes the pending state. It propagates naturally if you chain AsyncComputed task functions.

AsyncComputed doesn't add that to all signals because it's just a utility that runs on top of the current signals API without modifying it.

**Comment by eKoopmans** [comment]
Quick example of pending status not propagating:

```js
function logValueAndStatus(asyncComputed) {
	console.log(`${asyncComputed.status.padEnd(10)}${asyncComputed.get()}`);
}

const asyncComputed1 = new AsyncComputed(async() => {
	await new Promise(resolve => setTimeout(resolve, 1000));
	return 'async1';
});
const asyncComputed2 = new AsyncComputed(async() => {
	asyncComputed1.get();
	await new Promise(resolve => setTimeout(resolve, 500));
	return `${asyncComputed1.get()} async2`;
});

effect(() => logValueAndStatus(asyncComputed1));
effect(() => logValueAndStatus(asyncComputed2));
```

Above results in:
```
pending   undefined
pending   undefined
complete  undefined async2
complete  async1
pending   undefined async2
complete  async1 async2
```

Note the `complete  undefined async2` - the second async computed should not be "complete" while its dependency is still pending.

**Comment by justinfagnani** [comment]
Yes, you have to await something form the first task, like:

```ts
const asyncComputed2 = new AsyncComputed(async() => {
	const value1 = await asyncComputed1.complete;
	await new Promise(resolve => setTimeout(resolve, 500));
	return `${value1} async2`;
});
```

This follows normal async function / Promise chaining. This also propagates any errors.

**Comment by pzuraq** [comment]
Figure I might as well add my own stab at this since it’s live now, I also have async computeds, as well as subscriptions (my updated take on Relays): https://signalium.dev/

**Comment by eKoopmans** [comment]
@justinfagnani Thanks for the clarification! I don't think that's going to work for my needs, and to me breaks the signal pattern of getting synchronous updates about changing data. If nothing else, no regular `Computed` could ever rely on an `AsyncComputed` if it was necessary to await `.complete`.

I'll take a stab at a proposal, though it's possible that the limitations of the existing `State` and `Computed` implementations simply prevent what I'm trying to do.

**Comment by pzuraq** [comment]
@eKoopmans the way that regular computeds rely on async is that async computeds expose their result declaratively. So there is an `isLoading` property and a `result` property for instance.

This is much like handling Result or Option in functional languages with pattern matching. The sync computeds should handle the case where `isLoading` is true (e.g. show a spinner) and where it’s false and there’s a result.

Check out the design in Signalium, I wrote a whole doc explaining the nuances of this. I think this design space is fairly well understood at this point, so I’d look at prior art before coming up with a green field design.

**Comment by justinfagnani** [comment]
> [@justinfagnani](https://github.com/justinfagnani) Thanks for the clarification! I don't think that's going to work for my needs, and to me breaks the signal pattern of getting synchronous updates about changing data. If nothing else, no regular `Computed` could ever rely on an `AsyncComputed` if it was necessary to await `.complete`.

> [@eKoopmans](https://github.com/eKoopmans) the way that regular computeds rely on async is that async computeds expose their result declaratively. So there is an `isLoading` property and a `result` property for instance.

@pzuraq is right on here. It's not _necessary_ to await `.complete`, it's just that from within an async task function, awaiting propagates pending, result, and error states automatically. From within a regular computed you can manually propagate those things, like:

```ts
const syncComputed2 = new Signal.Computed(() => {
  if (asyncComputed1.status === 'pending') {
    return {status: 'pending'};
  }
  // this will throw if asyncComputed1.status === 'eror'
  const value1 = asyncComputed1.get();
  return {status: 'complete', value: value1};
});
```

So yes, this require manually wiring and propagating status through sync computeds, but again, this utility works on top of the proposal and doesn't attempt to modify it or the polyfill.

> I'll take a stab at a proposal, though it's possible that the limitations of the existing `State` and `Computed` implementations simply prevent what I'm trying to do.

I think that @dead-claudia is on the right track here for built-in async support, if it's feasible: extend the Computed state machine to allow the the `~computing~` state to persist until an async computation is completed, make a Computed's state `~computing~` if any of its dependencies are computing, and expose a portion of Computed's internal state as public fields (similar to AsyncComputed).

I still think that `await` is useful within async computeds because it lets you pause execution of the computed when you know you need to wait for a result. Without that you need to restart your computation every time a dependency resolves, similar to React hooks, and nicely highlighted as a problem in @pzuraq's Signalium post: https://signalium.dev/introduction#the-hooks-version Does that mean that all Computeds should have a promise that represents the async `~computing~` state? Not sure.

**Comment by ivan7237d** [comment]
Different angle on the isPending/cancellation discussion above: I built a lazy, cancelable, synchronously-settling single-shot Observable ([LazyPromise](https://github.com/lazy-promise/lazy-promise/tree/main/packages/core)) and used it as the basis for computed/effect in an [alien-signals integration](https://github.com/lazy-promise/lazy-promise/tree/main/packages/alien-signals). Subscribing is what starts the async work, unsubscribing cancels it — so a memo returning one of these gets caching, multicast, and cancel-on-recompute for free, without an extra field on every signal. There's also a small unbox() that exposes a plain T | undefined signal, undefined while pending.

Write-up + code in the README, in case the tradeoffs are useful here.
